File: StarcevMatvey/Term 4/ThreadPools/UnitTests/ThreadPoolTests.cs
using System;
using System.Threading;
using NUnit.Framework;
using ThreadPool;

namespace ThreadPooltests
{
    public class ThreadPoolTests
    {
        private static int i = 0;

        [Test]
        public void Test()
        {
            Action act = () => i++;

            var n = 4;
            var pool = new ThreadPool.ThreadPool(n);

            for(var i = 0; i < 10; i++) pool.Enqueue(act);
            Thread.Sleep(3000);

            pool.Dispose();

            Assert.AreEqual(i, 10);
        }

        [Test]
        public void StopTest()
        {
            Action act = () => i++;

            var n = 1;
            var pool = new ThreadPool.ThreadPool(n);

            pool.Dispose();

            Assert.Catch(() => pool.Enqueue(act));
        }
    }
}

File: VyacheslavBuchin/Term4/Block2/ThreadPool/ThreadPoolTests/ThreadPoolTests.fs
module ThreadPoolTests

open System
open System.Threading
open NUnit.Framework
open ThreadPool

[<Test>]
[<Repeat(5)>]
let ``Task are processed sequentially when thread is single`` () =
    let storage = ResizeArray()
    let task n () = lock storage (fun () -> storage.Add n)
    let work () =
        use pool = new ThreadPool(1, Seq.init 5 task)
        pool.Join()

    work()

    for i = 0 to storage.Count - 1 do
        Assert.AreEqual(i, storage[i])

[<Test>]
let ``Argument exception is raised when thread count is negative`` () =
    Assert.Throws(
        typeof<ArgumentException>,
        fun () ->
            use _ = new ThreadPool(-1)
            ()
    ) |> ignore

[<Test>]
[<Repeat(5)>]
let ``Enqueued tasks when joining is started should be also processed`` () =
    let storage = ResizeArray()
    let task _ () = lock storage (fun () -> storage.Add 1; Thread.Sleep(100))
    let work () =
        use pool = new ThreadPool(3, Seq.init 10 task)
        Thread(fun () -> Thread.Sleep(100); pool.Enqueue(task ())).Start()
        pool.Join()

    work()

    Assert.AreEqual(11, storage.Count)
